diff --git a/arch/mips/kernel/vmlinux.lds.S b/arch/mips/kernel/vmlinux.lds.S
index 1d5d021..e8c9c23 100644
--- a/arch/mips/kernel/vmlinux.lds.S
+++ b/arch/mips/kernel/vmlinux.lds.S
@@ -65,11 +65,6 @@ SECTIONS
 		*(.fixup)
 		*(.gnu.warning)
 	} :text = 0
-	. = ALIGN(0x4000);
-	__iram = .;
-	.iram : {
-		*(.iram)
-	}
 	_etext = .;	/* End of text section */
 
 	EXCEPTION_TABLE(16)
@@ -81,6 +76,24 @@ SECTIONS
 		__stop___dbe_table = .;
 	}
 
+	. = ALIGN(0x4000);
+	__iram = .;
+	__iram_start = . ;
+	.iram : {
+		*(.iram)
+	}
+	. = ALIGN(0x4000);
+	__iram_end = . ;
+
+	. = ALIGN(0x2000);
+	__dram = .;
+	__dram_start = .;
+	.dram : {
+		*(.dram)
+	}
+	. = ALIGN(0x2000);
+	__dram_end = .;
+
 #ifdef CONFIG_CAVIUM_OCTEON_SOC
 #define NOTES_HEADER
 #else /* CONFIG_CAVIUM_OCTEON_SOC */
@@ -104,14 +117,6 @@ SECTIONS
 		CONSTRUCTORS
 	}
 	BUG_TABLE
-	. = ALIGN(0x2000);
-	__dram = .;
-	__dram_start = .;
-	.dram : {
-		*(.dram)
-	}
-	. = ALIGN(0x2000);
-	__dram_end = .;
 	_gp = . + 0x8000;
 	.lit8 : {
 		*(.lit8)
