diff --git a/arch/mips/crt_arch.h b/arch/mips/crt_arch.h
index 9fc50d7..4285e63 100644
--- a/arch/mips/crt_arch.h
+++ b/arch/mips/crt_arch.h
@@ -16,11 +16,21 @@ __asm__(
 ".hidden _DYNAMIC \n"
 "	.gpword _DYNAMIC \n"
 "1:	lw $gp, 0($ra) \n"
+#ifndef __mlexra
 "	subu $gp, $ra, $gp \n"
 "	move $4, $sp \n"
+#else
+"	move $4, $sp \n"
+"	subu $gp, $ra, $gp \n"
+#endif /* !__mlexra */
 "	lw $5, 8($ra) \n"
+#ifndef __mlexra
 "	addu $5, $5, $gp \n"
 "	lw $25, 4($ra) \n"
+#else
+"	lw $25, 4($ra) \n"
+"	addu $5, $5, $gp \n"
+#endif /* !__mlexra */
 "	addu $25, $25, $gp \n"
 "	and $sp, $sp, -8 \n"
 "	jalr $25 \n"
diff --git a/arch/mips/reloc.h b/arch/mips/reloc.h
index b3d59a4..ef3be8e 100644
--- a/arch/mips/reloc.h
+++ b/arch/mips/reloc.h
@@ -36,6 +36,7 @@
 #define CRTJMP(pc,sp) __asm__ __volatile__( \
 	"move $sp,%1 ; jr %0" : : "r"(pc), "r"(sp) : "memory" )
 
+#ifndef __mlexra
 #define GETFUNCSYM(fp, sym, got) __asm__ ( \
 	".hidden " #sym "\n" \
 	".set push \n" \
@@ -50,3 +51,21 @@
 	"	addu %0, %0, $ra \n" \
 	".set pop \n" \
 	: "=r"(*(fp)) : : "memory", "ra" )
+#else
+#define GETFUNCSYM(fp, sym, got) __asm__ ( \
+	".hidden " #sym "\n" \
+	".set push \n" \
+	".set noreorder \n" \
+	"	bal 1f \n" \
+	"	 nop \n" \
+	"	.gpword . \n" \
+	"	.gpword " #sym " \n" \
+	"1:	lw %0, ($ra) \n" \
+	"	nop \n" \
+	"	subu %0, $ra, %0 \n" \
+	"	lw $ra, 4($ra) \n" \
+	"	nop \n" \
+	"	addu %0, %0, $ra \n" \
+	".set pop \n" \
+	: "=r"(*(fp)) : : "memory", "ra" )
+#endif /* !__mlexra */
diff --git a/crt/mips/crtn.s b/crt/mips/crtn.s
index 506a04b..9597196 100644
--- a/crt/mips/crtn.s
+++ b/crt/mips/crtn.s
@@ -3,11 +3,23 @@
 .section .init
 	lw $gp,24($sp)
 	lw $ra,28($sp)
+#ifndef __mlexra
 	j $ra
 	addu $sp,$sp,32
+#else /* __mlexra */
+	addu $sp,$sp,32
+	j $ra
+	nop
+#endif /* __mlexra */
 
 .section .fini
 	lw $gp,24($sp)
 	lw $ra,28($sp)
+#ifndef __mlexra
 	j $ra
 	addu $sp,$sp,32
+#else /* __mlexra */
+	addu $sp,$sp,32
+	j $ra
+	nop
+#endif /* __mlexra */
diff --git a/src/fenv/mips/fenv.S b/src/fenv/mips/fenv.S
index a5cb1f5..c62f299 100644
--- a/src/fenv/mips/fenv.S
+++ b/src/fenv/mips/fenv.S
@@ -64,6 +64,9 @@ fesetenv:
 	beq     $5, $0, 1f
 	 nop
 	lw      $5, 0($4)
+#ifdef __mlexra
+	nop
+#endif /* __mlexra */
 1:	ctc1    $5, $31
 	jr      $ra
 	li      $2, 0
diff --git a/src/ldso/mips/dlsym.s b/src/ldso/mips/dlsym.s
index 1573e51..8d5f544 100644
--- a/src/ldso/mips/dlsym.s
+++ b/src/ldso/mips/dlsym.s
@@ -13,5 +13,8 @@ dlsym:
 	jalr $25
 	nop
 	lw $ra, 12($sp)
+#ifdef __mlexra
+	nop
+#endif
 	jr $ra
 	addiu $sp, $sp, 16
diff --git a/src/setjmp/mips/longjmp.S b/src/setjmp/mips/longjmp.S
index fdb6c95..86bc832 100644
--- a/src/setjmp/mips/longjmp.S
+++ b/src/setjmp/mips/longjmp.S
@@ -36,5 +36,11 @@ longjmp:
 	lw      $22, 32($4)
 	lw      $23, 36($4)
 	lw      $30, 40($4)
+#ifndef __mlexra
 	jr      $ra
 	lw      $28, 44($4)
+#else
+	lw      $28, 44($4)
+	jr      $ra
+	nop
+#endif /* !__mlexra */
diff --git a/src/signal/mips/sigsetjmp.s b/src/signal/mips/sigsetjmp.s
index 74b65ff..de2bb77 100644
--- a/src/signal/mips/sigsetjmp.s
+++ b/src/signal/mips/sigsetjmp.s
@@ -15,6 +15,9 @@ __sigsetjmp:
 	sw $16, 104+4+16($4)
 
 	lw $25, %call16(setjmp)($gp)
+#ifdef __mlexra
+	nop
+#endif /* __mlexra */
 	jalr $25
 	 move $16, $4
 
@@ -25,9 +28,15 @@ __sigsetjmp:
 
 .hidden __sigsetjmp_tail
 	lw $25, %call16(__sigsetjmp_tail)($gp)
+#ifdef __mlexra
+	nop
+#endif /* __mlexra */
 	jr $25
 	 nop
 
 1:	lw $25, %call16(setjmp)($gp)
+#ifdef __mlexra
+	nop
+#endif /* __mlexra */
 	jr $25
 	 nop
diff --git a/src/thread/mips/syscall_cp.s b/src/thread/mips/syscall_cp.s
index d284626..e07fd4e 100644
--- a/src/thread/mips/syscall_cp.s
+++ b/src/thread/mips/syscall_cp.s
@@ -17,6 +17,9 @@ __syscall_cp_asm:
 	subu    $sp, $sp, 32
 __cp_begin:
 	lw      $4, 0($4)
+#ifdef __mlexra
+	nop
+#endif /* __mlexra */
 	bne     $4, $0, __cp_cancel
 	move    $2, $5
 	move    $4, $6
@@ -46,8 +49,13 @@ __cp_cancel:
 	.gpword .
 	.gpword __cancel
 1:	lw      $3, ($ra)
+#ifndef __mlexra
 	subu    $3, $ra, $3
 	lw      $25, 4($ra)
+#else
+	lw      $25, 4($ra)
+	subu    $3, $ra, $3
+#endif /* !__mlexra */
 	addu    $25, $25, $3
 	jr      $25
 	move    $ra, $2
diff --git a/src/unistd/mips/pipe.s b/src/unistd/mips/pipe.s
index ba2c39a..ed10f9a 100644
--- a/src/unistd/mips/pipe.s
+++ b/src/unistd/mips/pipe.s
@@ -11,6 +11,9 @@ pipe:
 	beq $7, $0, 1f
 	nop
 	lw $25, %call16(__syscall_ret)($gp)
+#ifdef __mlexra
+	nop
+#endif /* __mlexra */
 	jr $25
 	subu $4, $0, $2
 1:	sw $2, 0($4)
